index.rst

```diff
@@ -13,17 +13,256 @@
+The current boundaries between the Butler and its Registry and Datastore components are under strain in a number of different ways.
+Failure recovery during deletion operations has long been in bad shape, and much of the current "trash"-based system is currently just unused complexity.
+Butler client/server will require new approaches to atomic operations and managing operation latency (including caching), and :jira:`RFC-888` has recently shown that we may want to move away from the Registry component providing public APIs even outside of the client/server.
+The provenance system proposed in :cite:`DMTN-205` can also impact these boundaries, especially if we want to extend its notion of QuantumGraph storage in data repositories to graphs that have not yet been fully executed; one possibility here is to add a third component that acts as a more graph-oriented Registry.
+This technote will propose a new high-level organization of Butler interfaces and responsibilities to address these concerns.
 
-Add content here
+Consistency across Registry and Datastore
+=========================================
+
+Current defects
+---------------
+
+The current boundary between Registry and Datastore was set up with two principles in mind:
+
+- a dataset has to be added to Registry first, so it can take responsibility for generating a (at the time autoincrement integer) unique dataset ID;
+- we should use database transactions that are not committed until Datastore operations are completed to maintain consistency across the two components.
+
+The first is no longer true - we've switched to UUIDs instead - specifically to support writing to Datastore first, via BPS database-free schemes like execution butler (see :cite:`DMTN-177`).
+Using database transactions has never really worked for deletes, fundamentally because Datastore operations cannot reliably be rolled back if the database commit itself fails.
+Our attempt to work around this with a system that moves datasets to a "trash" table had considerable problems of its own, leaving us with no real attempt to maintain integrity between Datastore and Registry.
+Even for ``put`` calls, starting a transaction before the Datastore write begins is problematic because it keeps database transactions open longer than we'd like.
+
+High-level proposal
+-------------------
+
+I propose we adopt the following consistency principles instead.
+
+1. A dataset can be present in either Registry or Datastore without being present in the other.
+
+2. A dataset present in Datastore alone must have a data ID that is valid in the Registry for that data repository (i.e. it uses valid dimension values) and it must not have any Datastore records in the Registry database.
```

Comment:
Writing the predicted output records to the graph might work. It might also be that a standard Butler with a graph-backed Registry object would work in the new Butler design.

We could think about a `datastore.put()` that takes optional records and then ignores all the file template and formatter selection configuration but it would till have to know about formatter write parameters (such as data compression settings) so I don't think it's possible to get away from any datastore configuration -- unless you put that in the graph as well and then have that be passed into the Datastore when Butler is created.